dbms/src/Storages/DeltaMerge/File/DMFileBlockInputStream.cpp

```diff
@@ -140,19 +140,19 @@ SkippableBlockInputStreamPtr DMFileBlockInputStreamBuilder::tryBuildWithVectorIn
         return build(dmfile, read_columns, rowkey_ranges, scan_context);
     };
 
-    if (rs_filter == nullptr)
+    if (!rs_filter)
         return fallback();
 
     // Fast Scan and Clean Read does not affect our behavior. (TODO(vector-index): Confirm?)
     // if (is_fast_scan || enable_del_clean_read || enable_handle_clean_read)
     //    return fallback();
 
     auto filter_with_ann = std::dynamic_pointer_cast<WithANNQueryInfo>(rs_filter);
-    if (filter_with_ann == nullptr)
+    if (!filter_with_ann)
         return fallback();
 
     auto ann_query_info = filter_with_ann->ann_query_info;
-    if (!ann_query_info || ann_query_info->top_k() == std::numeric_limits<UInt32>::max())
```

Comment:
I think the behavior of when `ann_query_info->top_k() == std::numeric_limits<UInt32>::max()` fallback to "brute force" algo is missed after these changes?

Reply:
https://github.com/pingcap/tidb/pull/57307 Maybe handling this behavior in TiDB is a better idea.

Reply:
As the terms of "defensive programming", I think both TiDB and TiFlash should try to prevent top_k() == UINT32_MAX for running into vector index. Because in TiFlash perspective, we can not ensure the input is always trustable, especially the "top_k" value is comes from a network package. Maybe the network packet comes from a lower version of tidb-server, cse version of tidb-server, or other clients. Or in the TiDB side, someone may remove the limitation you added in someday.
So I still think the guard in TiFlash side is necessary.

Reply:
```cpp
     bool is_valid_ann_query = ann_query_info->top_k() != std::numeric_limits<UInt32>::max();
     bool is_matching_ann_query = std::any_of(
         table_column_defines.begin(),
         table_column_defines.end(),
         [cid = ann_query_info->column_id()](const ColumnDefine & cd) -> bool { return cd.id == cid; });
     if (!is_valid_ann_query || !is_matching_ann_query)
         return rs_operator;

     return wrapWithANNQueryInfo(rs_operator, ann_query_info);
```